File: clients/go/model/mine_state.go
package model

import "io"
import . "aicup2019/stream"

type MineState int32
const (
    MineStatePreparing MineState = 0
    MineStateIdle MineState = 1
    MineStateTriggered MineState = 2
    MineStateExploded MineState = 3
)
func ReadMineState(reader io.Reader) MineState {
    switch ReadInt32(reader) {
    case 0:
        return MineStatePreparing
    case 1:
        return MineStateIdle
    case 2:
        return MineStateTriggered
    case 3:
        return MineStateExploded
    }
    panic("Unexpected discriminant value")
}

File: clients/javascript/model/mine-state.js
class MineState {
    constructor (discriminant) {
        this.discriminant = discriminant;
    }

    static async readFrom (stream) {
        switch (await stream.readInt()) {
        case 0:
            return new MineState(MineState.PREPARING);
        case 1:
            return new MineState(MineState.IDLE);
        case 2:
            return new MineState(MineState.TRIGGERED);
        case 3:
            return new MineState(MineState.EXPLODED);
        default:
            throw new Error('Unexpected discriminant value');
        }
    }
}

MineState.PREPARING = 0;
MineState.IDLE = 1;
MineState.TRIGGERED = 2;
MineState.EXPLODED = 3;

module.exports.MineState = MineState;
